Add plexon 2 support

h-mayorquin opened the pull request:
I am able to run local conversions, I need to add tests to the ci.

Should fix https://github.com/catalystneuro/neuroconv/issues/204

To-do:

- [ ] Set up CI tests that take into account wine in ubuntu
- [ ] See if I get windows error on the CI, fix it.
- [ ] Set up metadata from the annotations. See below.
- [ ] Add example to conversion gallery.
- [ ] Figure out if stream should be blocked or accessible. Making it accessible but setting default as spikeinterface now. See header for signals below.

Metadata available:

```
'name': 'Block containing PL2 data#0',
 'm_CreatorComment': b'4ch sorted spikes, continuous wideband, and LFP. Events on event channels 1-16, strobe values 32513-32767. One aux input.',
 'm_CreatorSoftwareName': b'OmniPlex',
 'm_CreatorSoftwareVersion': b'1.10.29',
 'm_CreatorDateTime': datetime.datetime(113, 11, 20, 15, 59, 39),
 'm_CreatorDateTimeMilliseconds': 0,
 'm_TimestampFrequency': 40000.0,
 'm_NumberOfChannelHeaders': 334,
 'm_TotalNumberOfSpikeChannels': 64,
 'm_NumberOfRecordedSpikeChannels': 64,
 'm_TotalNumberOfAnalogChannels': 224,
 'm_NumberOFRecordedAnalogChannels': 69,
 'm_NumberOfDigitalChannels': 46,
 'm_MinimumTrodality': 1,
 'm_MaximumTrodality': 1,
 'm_NumberOfNonOmniPlexSources': 0,
 'm_Unused': 0,
 'm_ReprocessorComment': b'',
 'm_ReprocessorSoftwareName': b'',
 'm_ReprocessorSoftwareVersion': b'',
 'm_ReprocessorDateTime': None,
 'm_ReprocessorDateTimeMilliseconds': 0,
 'm_StartRecordingTime': 4470476,
 'm_DurationOfRecording': 1179268}
```

The header for the signals:

```python
{'nb_block': 1,
 'nb_segment': [1],
 'signal_streams': array([('stream@40000.0Hz', '3'), ('stream@40000.0Hz', '4'),
        ('stream@1000.0Hz', '7'), ('stream@1000.0Hz', '12')],
       dtype=[('name', '<U64'), ('id', '<U64')]),
 'signal_channels': array([('WB01', 'source3.1', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '3'),
        ('WB02', 'source3.2', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '3'),
        ('WB03', 'source3.3', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '3'),
        ('WB04', 'source3.4', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '3'),
        ('SPKC01', 'source4.1', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '4'),
        ('SPKC02', 'source4.2', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '4'),
        ('SPKC03', 'source4.3', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '4'),
        ('SPKC04', 'source4.4', 40000., 'int16', 'Volts', 6.10351563e-07, 0., '4'),
        ('FP01', 'source7.1',  1000., 'int16', 'Volts', 6.10351563e-07, 0., '7'),
        ('FP02', 'source7.2',  1000., 'int16', 'Volts', 6.10351563e-07, 0., '7'),
        ('FP03', 'source7.3',  1000., 'int16', 'Volts', 6.10351563e-07, 0., '7'),
        ('FP04', 'source7.4',  1000., 'int16', 'Volts', 6.10351563e-07, 0., '7'),
        ('AI01', 'source12.1',  1000., 'int16', 'Volts', 1.52587891e-04, 0., '12')],
       dtype=[('name', '<U64'), ('id', '<U64'), ('sampling_rate', '<f8'), ('dtype', '<U16'), ('units', '<U64'), ('gain', '<f8'), ('offset', '<f8'), ('stream_id', '<U64')]),
 'spike_channels': array([('SPK01.0', 'unit1.0', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK01.1', 'unit1.1', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK01.2', 'unit1.2', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK01.3', 'unit1.3', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK01.4', 'unit1.4', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK02.0', 'unit2.0', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK02.1', 'unit2.1', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK02.2', 'unit2.2', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK02.3', 'unit2.3', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK02.4', 'unit2.4', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK03.0', 'unit3.0', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK03.1', 'unit3.1', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK03.2', 'unit3.2', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK03.3', 'unit3.3', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK03.4', 'unit3.4', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK04.0', 'unit4.0', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK04.1', 'unit4.1', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK04.2', 'unit4.2', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK04.3', 'unit4.3', 'Volts', 6.10351563e-07, 0., 8, 40000.),
        ('SPK04.4', 'unit4.4', 'Volts', 6.10351563e-07, 0., 8, 40000.)],
       dtype=[('name', '<U64'), ('id', '<U64'), ('wf_units', '<U64'), ('wf_gain', '<f8'), ('wf_offset', '<f8'), ('wf_left_sweep', '<i8'), ('wf_sampling_rate', '<f8')]),
 'event_channels': array([('KBD1', '1', b'event'), ('KBD2', '2', b'event'),
        ('KBD3', '3', b'event'), ('KBD4', '4', b'event'),
        ('KBD5', '5', b'event'), ('KBD6', '6', b'event'),
        ('KBD7', '7', b'event'), ('KBD8', '8', b'event'),
        ('EVT01', '1', b'event'), ('EVT02', '2', b'event'),
        ('EVT03', '3', b'event'), ('EVT04', '4', b'event'),
        ('EVT05', '5', b'event'), ('EVT06', '6', b'event'),
        ('EVT07', '7', b'event'), ('EVT08', '8', b'event'),
        ('EVT09', '9', b'event'), ('EVT10', '10', b'event'),
        ('EVT11', '11', b'event'), ('EVT12', '12', b'event'),
        ('EVT13', '13', b'event'), ('EVT14', '14', b'event'),
        ('EVT15', '15', b'event'), ('EVT16', '16', b'event'),
        ('EVT17', '17', b'event'), ('EVT18', '18', b'event'),
        ('EVT19', '19', b'event'), ('EVT20', '20', b'event'),
        ('EVT21', '21', b'event'), ('EVT22', '22', b'event'),
        ('EVT23', '23', b'event'), ('EVT24', '24', b'event'),
        ('EVT25', '25', b'event'), ('EVT26', '26', b'event'),
        ('EVT27', '27', b'event'), ('EVT28', '28', b'event'),
        ('EVT29', '29', b'event'), ('EVT30', '30', b'event'),
        ('EVT31', '31', b'event'), ('EVT32', '32', b'event'),
        ('Strobed', '1', b'event'), ('RSTART', '2', b'event'),
        ('RSTOP', '3', b'event'), ('CPX1', '1', b'event'),
        ('CPX2', '2', b'event'), ('CPX3', '3', b'event')],
       dtype=[('name', '<U64'), ('id', '<U64'), ('type', 'S5')])}
```

bendichter commented:
@h-mayorquin see https://github.com/NeuralEnsemble/python-neo/blob/1a7f60e2d75176b6df00fd7a2d1344772d223174/.github/workflows/io-test.yml#L104-L109 for how @JuliaSprenger did this in neo

bendichter commented on src/neuroconv/datainterfaces/ecephys/plexon/plexondatainterface.py:
```diff
@@ -48,6 +48,49 @@ def get_metadata(self) -> DeepDict:
+            stream_id=stream_id,
+        )
+
+    def get_metadata(self) -> DeepDict:
+        metadata = super().get_metadata()
+
+        return metadata
```
```suggestion
```

there is no need to override this if you are not changing the metadata at all

h-mayorquin replied:
Yes, but I might, it looks liker there is session data in the annotations.  I just wanted to get it running quickly, I will feel this with some of the annotations available!

bendichter commented:
@h-mayorquin 
```
================================== FAILURES ===================================
____________________________ [doctest] plexon2.rst ____________________________
012 .. code-block:: python
013 
014     >>> from datetime import datetime
015     >>> from zoneinfo import ZoneInfo
016     >>> from pathlib import Path
017     >>> from neuroconv.datainterfaces import Plexon2RecordingInterface
018     >>>
019     >>> file_path = f"{ECEPHY_DATA_PATH}/plexon/4chDemoPL2.pl2"
020     >>> # Change the file_path to the location in your system
021     >>> interface = Plexon2RecordingInterface(file_path=file_path, verbose=False)
Expected nothing
Got:
    Downloading plexon dll to C:\Users\runneradmin\.plexon_dlls_for_neo\PL2FileReader64.dll
```

It looks like this failure is due to the fact that the gallery expected no string print out and neo automatically prints out this string informing the user that it is downloading the Plexon DLL. I guess you could either add that to the snippet or we could find a way to prevent the print-out.

h-mayorquin commented:
Yep, that's why I am against generating standard output by default. I think we have dealt with this before at some point. Let me think through it. The problem is that if you run the test for the first time the file gets downloaded and you get the message, but then, the next time you run the test (for example locally) there is no message (because the file is already there) and the test will fail if it was expecting that message.

h-mayorquin commented:
Hell yeah, the mac tests are passing now with the patch on https://github.com/NeuralEnsemble/python-neo/pull/1494.

Now we need to figure out the metadata and the random issue with the file handle and this should be ready!

bendichter commented:
wow, fully passing! Amazing, @h-mayorquin !

Let's work on getting this PR merged in neo

bendichter commented:
Also, this is being testing in the gallery but not in the unit tests. It would be nice to set some up so we can ensure that the metadata is being extracted accurately.

h-mayorquin commented:
> There's clearly a lot of metadata we could pull here. At a bare minimum, let's take the field m_CreatorDateTime and set it to the session_start_time.

Yes, this is done. I just needed to confirm that this was 1900 based and I changed that at neo so it is corrected from the base:

https://github.com/NeuralEnsemble/python-neo/pull/1495#issuecomment-2184196635

I added the metadata test for the datetime.

Let's have a brief discussion in our next meeting on how to add / interpret the rest of the metadata because I am at a loss with it.

h-mayorquin commented on tests/test_on_data/test_gin_ecephys/test_raw_recordings.py:
```diff
@@ -95,6 +96,12 @@ class TestEcephysRawRecordingsNwbConversions(unittest.TestCase):
             ),
             case_name="plexon_recording",
         ),
+        param(
```
This is the unit test for the conversion @bendichter

h-mayorquin commented on tests/test_on_data/test_metadata/test_plexon_metadata.py:
and here the metadata test.

h-mayorquin commented:
Ok, so the PRs are merge on neo side. We will have to wait for a release on their side. The test will be failing because of a caching issue with the test datasets I think. I will fix this once we have release but this should be working.

bendichter commented on .github/workflows/dev-testing.yml:
```diff
@@ -70,7 +75,7 @@ jobs:
         with:
           path: ./ephy_testing_data
-          key: ephys-datasets-2024-03-27-${{ matrix.os }}-${{ steps.ephys.outputs.HASH_EPHY_DATASET }}
+          key: ephys-datasets-2024-06-21-ubuntu-latest-${{ steps.ephys.outputs.HASH_EPHY_DATASET }}
```
What's going on with this change. Had we been storing 3 copies of the same data in the cache?

h-mayorquin replied:
This action only runs in ubuntu-latest:

https://github.com/catalystneuro/neuroconv/blob/cb8687f6456dea51545144bb2e81a0ee4fe36c02/.github/workflows/dev-testing.yml#L20-L22

{{matrix.os}} is not defined here and it was working because the cache action has implicit mechanism to match close strings I think.

I made it explicit to avoid any surprises.

bendichter commented:
This all looks good. Just one question about data caching.

h-mayorquin commented:
We can offer support now, yei!
